api/src/test/java/org/openmrs/api/EncounterServiceTest.java

```diff
@@ -98,7 +99,127 @@
+		ObsService os = Context.getObsService();
+
+		Encounter enc = es.getEncounter(100);
+
+		Obs o = os.getObs(101);
+		o.setValueText("Obs value updated");
+
+		es.saveEncounter(enc);
+		Context.evictFromSession(enc);
+		Context.flushSession();
+
+		enc = es.getEncounter(100);
+
+		Set<Obs> obsAtTopLevelUpdated = enc.getObsAtTopLevel(true);
+		Obs oParent = os.getObs(100);
+		final Obs editedObs = os.getObs(101);
+		Obs o2 = os.getObs(102);
+		Obs o3 = getNewVersionOfEditedObs(oParent,editedObs);
+
+		assertEquals(1,obsAtTopLevelUpdated.size());
+		assertEquals(3, oParent.getGroupMembers(true).size());
+		assertTrue(editedObs.getVoided());
+		assertFalse(oParent.getVoided());
+		assertFalse(o2.getVoided());
+
+		assertNotNull(o3);
+		assertFalse(o3.getVoided());
+		assertEquals("Obs value updated", o3.getValueText());
+	}
+
+	@Test
+	@Verifies(value = "should update existing encounter correctly when a new obs is added to parent obs", method = "saveEncounter(Encounter)")
+	public void saveEncounter_shouldUpdateExistingEncounterWhenNewObsIsAddedToParentObs() throws Exception {
+		executeDataSet(ENC_OBS_HIERARCHY_DATA_XML);
+		ConceptService cs = Context.getConceptService();
+		EncounterService es = Context.getEncounterService();
+		ObsService os = Context.getObsService();
+
+		Encounter enc = es.getEncounter(100);
+
+		Obs o3 = new Obs();
+		o3.setConcept(cs.getConcept(3));
+		o3.setDateCreated(new Date());
+		o3.setCreator(Context.getAuthenticatedUser());
+		o3.setLocation(new Location(1));
+		o3.setObsDatetime(new Date());
+		o3.setPerson(Context.getPersonService().getPerson(3));
+		o3.setValueText("third obs value text");
+		o3.setEncounter(enc);
+
+		Obs oParent = os.getObs(100);
+		oParent.addGroupMember(o3);
+
+		es.saveEncounter(enc);
+
+		Context.evictFromSession(enc);
+		Context.flushSession();
+
+		enc = es.getEncounter(100);
+
+		Set<Obs> obsAtTopLevelUpdated = enc.getObsAtTopLevel(true);
+		assertEquals(1,obsAtTopLevelUpdated.size());
+
+		assertEquals(3, obsAtTopLevelUpdated.iterator().next().getGroupMembers(true).size());
+
+		oParent = os.getObs(100);
+		assertTrue(oParent.getGroupMembers(true).contains(os.getObs(101)));
+		assertTrue(oParent.getGroupMembers(true).contains(os.getObs(102)));
+		assertTrue(oParent.getGroupMembers(true).contains(os.getObs(o3.getObsId())));
+	}
+
+
+	@Test
+	@Verifies(value = "should update existing encounter correctly when an existing obs is removed from parent obs", method = "saveEncounter(Encounter)")
+	public void saveEncounter_shouldSaveEncounterWhenTopLevelObsIsUpdatedByRemovingChildObs() throws Exception {
+		executeDataSet(ENC_OBS_HIERARCHY_DATA_XML);
+		EncounterService es = Context.getEncounterService();
+		ObsService os = Context.getObsService();
+
+		Encounter enc = es.getEncounter(100);
+		Obs oParent = os.getObs(100);
+		Obs oChild = os.getObs(101);
+
+		oParent.removeGroupMember(oChild);
+
+		es.saveEncounter(enc);
+
+		Context.evictFromSession(enc);
```

Comment:
@wluyima I agree with you and I see this test case fails.  The reason it fails as follows: When the original oChild is [removed](https://github.com/bharatak/openmrs-core/blob/76ec6f3b4d6a6ed39de892cfda35e7f25efc144c/api/src/test/java/org/openmrs/api/EncounterServiceTest.java#L191) from oParent, the obsGroup is set to null in oChild.  But while we save the oChild, we evict the original obs to get the fresh one [here](https://github.com/bharatak/openmrs-core/blob/76ec6f3b4d6a6ed39de892cfda35e7f25efc144c/api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java#L131) and there by resetting the obs_group to previous parent.